### PR TITLE
PR2-Feature changes to support hash verification, file resume, drive…

### PR DESCRIPTION
**Pull request by Arvinds-ds**
New features

a) Better progress update giving ETA/Speed
b) Human readable size and time
c) pause/resume downloads with resume=True option
d) hash verification if hash is provided (hash_true option)
e) Special handling for google drive downloads

### CODE REVIEW

File: observations/util.py [hunk 1]
```diff
@@ -3,15 +3,527 @@
...
+  print("Supports File Size:   [{}] ".format('Y'
+                                             if cap_dict['supports_file_size']
+                                             else 'N'))
+  print("File Size:            [{}] ".format(cap_dict['file_size']))
+  print("Additional GET params [{}]".format(cap_dict['addnl_params']))
+
+
+def check_capabilities(url, start=120, num_bytes=300, user_agent=None,
+                       _print=True):
+  """ Check capabilities of a specified URL. This function queries the URL
```
Comment (dustinvtran) [inline observations/util.py:88]: Remove leading space.

File: observations/util.py [hunk 2]
```diff
@@ -3,15 +3,527 @@
...
 import tarfile
+import time
 import zipfile
 
 from six.moves import urllib
 
 
-def maybe_download_and_extract(directory, url, extract=True):
+def _get_google_confirm_id_token(url, session):
+  """Helper function to get confirmation token for google drive
```
Comment (dustinvtran) [inline observations/util.py:20]: Add period at end of all first docstring sentences.

File: observations/util.py [hunk 3]
```diff
@@ -3,15 +3,527 @@
...
+    for key, value in response.cookies.items():
+      if key.startswith('download_warning'):
+        return file_id, value
+    return file_id, None
+  except KeyError:
+    return None, None
+
+
+def _gdrive_params(url, session):
+  """Helper function reurn google drive specific parameters
```
Comment (dustinvtran) [inline observations/util.py:44]: `Helper function returning google drive specific parameters.`

File: observations/util.py [hunk 4]
```diff
@@ -3,15 +3,527 @@
...
+                                             if cap_dict['supports_file_size']
+                                             else 'N'))
+  print("File Size:            [{}] ".format(cap_dict['file_size']))
+  print("Additional GET params [{}]".format(cap_dict['addnl_params']))
+
+
+def check_capabilities(url, start=120, num_bytes=300, user_agent=None,
+                       _print=True):
+  """ Check capabilities of a specified URL. This function queries the URL
+  to get a response on the remote server capabilities.Requests for num_bytes
```
Comment (dustinvtran) [inline observations/util.py:89]: `. Requests`

File: observations/util.py [hunk 5]
```diff
@@ -3,15 +3,527 @@
...
+                       'session': session
+                       }
+  if _print is True:
+    _print_capabilities(capabilities_dict)
+  return capabilities_dict
+
+
+def humanize_size(n):
+  """Convert file size in bytes to a friendly human readable form.
+  E.g humanize_size(1024) - '1 KB'.
```
Comment (dustinvtran) [inline observations/util.py:152]: Examples are written at end of docstring like:
```
#### Examples

```python
humanize_size(1024)
## '1 KB'
```

See, e.g., https://github.com/blei-lab/edward/blob/master/edward/inferences/inference.py.

File: observations/util.py [hunk 6]
```diff
@@ -3,15 +3,527 @@
...
+  else:
+    return False
+
+
+def check_file_downloaded(file_path, file_size):
+  """Check whether a file of specified size exists.
+
+    Args:
+      file_path: str.
+        Full path of the file.
```
Comment (dustinvtran) [inline observations/util.py:292]: `file_path` vs `path` seems inconsistent; should stick with `path`.

File: observations/util.py [hunk 7]
```diff
@@ -3,15 +3,527 @@
...
+  local_file_size = os.path.getsize(file_path) if file_exists else -1
+  if file_size == local_file_size:
+    return True
+  else:
+    return False
+
+
+def _print_progress(file_path, dl_bytes, file_size, block_size, count,
+                    _start_time):
+  """ Helper function that is called to print progress of download.
```
Comment (dustinvtran) [inline observations/util.py:309]: `"""Helper function`

File: observations/util.py [hunk 8]
```diff
@@ -3,15 +3,527 @@
...
+        Bytes already downloaded in bytes.
+      file_size. int.
+        Total file size in bytes.
+      block_size: int.
+        Block size of bytes requested per iteration.
+      count: int.
+        Iteration count.
+      _start_time: object.
+        Time object to indicate when download started.
+    Returns:
```
Comment (dustinvtran) [inline observations/util.py:324]:
```
      Time object ...

  Returns:
```

File: observations/util.py [hunk 9]
```diff
@@ -3,15 +3,527 @@
...
+      User agent which is sometimes required to query some domains for proper
+      download.
+    _print: bool, optional.
+       Print the remote server capabilities.
+
+  Returns
+    dict: Dictionary of remote server capabilities, required prior to full
+          download.
+  """
+  supports_range = False
```
Comment (dustinvtran) [inline observations/util.py:110]: Should `supports_range` be called `supports_resume`?

File: observations/util.py [hunk 10]
```diff
@@ -3,15 +3,527 @@
...
+                     'Unknown',
+                     humanize_size(speed),
+                     humanize_time(duration)))
+    sys.stdout.flush()
+
+
+def normal_download(url, file_path, session, params={}, headers={},
+                    hash_true=None, timeout=10, block_size=1024 * 1024):
+  """start download of a file. No pause/resume capability.
+    Args:
```
Comment (dustinvtran) [inline observations/util.py:360]: `Start download...`. blankline between all sections

**Review by dustinvtran — COMMENTED** [review]
Excellent work as mentioned before. I only have a bunch of minor comments. What I'll do is merge it now and submit it a new PR making the minor changes and ask you for feedback.

**Comment by Arvinds-ds** [comment]
Pls go ahead with your changes. Happy to contribute.

**Pull request merged**